content/blog-posts/2025-04-01-tolkein-ring/index.md

```diff
@@ -0,0 +1,46 @@
+ogImage: tolkien-ebee.png
+toc: false
+categories:
+  - Community
+---
+
+![A bee with a ring...for now](tolkien-ebee.png)
+
+Author: Bill Mulligan - Isovalent
+
+Before Wifi there was Ethernet. Before Ethernet there was [Token Ring](https://en.wikipedia.org/wiki/Token_Ring). And before Token Ring there was Tolkien Ring. Before Tolkien Ring, the packets wandered without purpose, lost to the void.
+
+Now support for Tolkien Ring is coming to eBPF. Since its humble beginnings as a packet processor, eBPF has expanded to become a general purpose runtime enabling developers to extend and rebuild whole swaths of the kernel better from low overhead observability to granular security, optimized scheduling, and continuous profiling. eBPF is now returning to its roots with its most ambitious project yet, native support for the original networking protocol Tolkien Ring.
+
+For those unfamiliar, Tolkien Ring is the oldest networking protocol, established in the Days of Númenor, when the world was young and packets were not yet buried under layers of abstraction. It was the first fully decentralized, peer-to-peer network where messages were passed by whispers on the wind, encoded in runes, and secured by the strength of oaths. But as time passed, its wisdom was lost, replaced by lesser protocols constrained by mere physical wires and fleeting radio waves, forgotten by all but the wisest maintainers. Now, thanks to eBPF, this ancient technology is being restored to its former glory.
+
+“eBPF already transformed networking at Meta twice, first with [Katran](https://engineering.fb.com/2018/05/22/open-source/open-sourcing-katran-a-scalable-network-load-balancer/) processing millions of packets of cat pictures per second then again with [netkit making container networking](https://fosdem.org/2025/schedule/event/fosdem-2025-4045-an-introduction-to-netkit-the-bpf-programmable-network-device/) with isolation from orcs as fast as host networking.” We’ve also seen eBPF expand into networking use cases we never imagined like HTTP parsing in the kernel or a [Layer 2 switch in Kubernetes for VMs](https://github.com/kube-vm-project/eSwitch), but bringing back Tolkien Ring was a challenge unlike any other. Half the documentation was in Elvish and the other half was a subdialect of Dwarvish last recorded in the times of Moria’s First Age maintainers. “Adapting Tolkien Ring to modern infrastructure required pushing eBPF to its limits, but in the end, it was the only technology capable of binding them all.” said Alexei Starovoitov eBPF co-founder and co-maintainer.
+
+Integrating Tolkien Ring into the kernel came with a few surprises. For example, Tolkien Ring does not retransmit packets, it simply ensures they arrive precisely when they are meant to. Packet loss is not a possibility but a matter of destiny, and unlike conventional cryptographic handshakes, Tolkien Ring’s security model is built on binding oaths, which once sworn, cannot be revoked.
```

Comment:
```suggestion
Integrating Tolkien Ring into the kernel came with a few surprises. For example, Tolkien Ring does not retransmit packets, it simply ensures they arrive precisely when they are meant to. Packet loss is not a possibility but a matter of destiny, and unlike conventional cryptographic handshakes, Tolkien Ring’s security model is built on binding oaths, which once sworn, cannot be revoked. eBPF maintainers also had to fight hard against the Tolkien Ring Wraiths: the Nine were once features, but time and corruption had turned them into CI flakes. Happily, all fell back into order after a few additional patches, which saved everybody a trip to Mount Doom.
```

Just an idea. Feel free to adjust as you see fit, or discard entirely.

Reply:
haha love it!